### PR TITLE
Add rate limiting DoFn

### PR DESCRIPTION
**Pull request by rgruener**
@nevillelyh @regadas @andrewsmartin

### CODE REVIEW

File: scio-extra/src/test/scala/com/spotify/scio/extra/transforms/RateLimiterDoFnTest.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.concurrent.TimeUnit
+
+import com.spotify.scio.testing._
+
+class RateLimiterDoFnTest extends PipelineSpec {
+
+  private val input = 0 until 10
+
+  it should "work" in {
```
Comment (nevillelyh) [inline scio-extra/src/test/scala/com/spotify/scio/extra/transforms/RateLimiterDoFnTest.scala]: `"RateLimiterDoFn"`

File: scio-extra/src/main/scala/com/spotify/scio/extra/transforms/package.scala [hunk 2]
```diff
@@ -112,4 +115,48 @@ package object transforms {
...
+     *                             per second
+     */
+    def withRateLimit(maxElementsPerSecond: Double): SCollection[T] = {
+      val runner = self.context.pipeline.getRunner
+      val maxNumWorkers = runner match {
+        case _: DirectRunner => 1
+        case _: DataflowRunner =>
+          val pipelineOptions = self.context.optionsAs[DataflowPipelineOptions]
+          val numWorkers = Math.max(pipelineOptions.getNumWorkers, pipelineOptions.getMaxNumWorkers)
+          if (numWorkers == 0) {
```
Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/transforms/package.scala]: Use `require()`?

File: scio-extra/src/main/scala/com/spotify/scio/extra/transforms/package.scala [hunk 3]
```diff
@@ -112,4 +115,48 @@ package object transforms {
...
+            throw new RuntimeException(
+              "Rate limiting only available when numWorkers or maxNumWorkers is explicitly set"
+            )
+          }
+          numWorkers
+        case _ => throw new NotImplementedError(
+          s"rateLimitThroughput not implemented for runner ${runner.getClass.toString}"
```
Comment (nevillelyh) [inline scio-extra/src/main/scala/com/spotify/scio/extra/transforms/package.scala]: Minor: `.toString` is redundant.

**Comment by nevillelyh** [comment]
Next time prefix branch name with username? Easy to keep track of things for long living branches.

**Comment by regadas** [comment]
👍 :shipit:

**Comment by rgruener** [comment]
:+1: will prefix next time

**Pull request merged**